File: src/commands/call.ts
/**
 * Call command - Execute a tool with arguments
 *
 * Output behavior:
 * - Default: Raw text content to stdout (CLI-friendly)
 * - With --json: Full JSON response to stdout
 * - Errors always go to stderr
 */

import {
  type McpConnection,
  debug,
  getConnection,
  getTimeoutMs,
  safeClose,
} from '../client.js';
import {
  type McpServersConfig,
  type ServerConfig,
  getServerConfig,
  loadConfig,
} from '../config.js';
import {
  ErrorCode,
  formatCliError,
  invalidJsonArgsError,
  invalidTargetError,
  serverConnectionError,
  toolExecutionError,
  toolNotFoundError,
} from '../errors.js';
import { formatJson, formatToolResult } from '../output.js';

export interface CallOptions {
  target: string; // "server/tool"
  args?: string; // JSON arguments
  configPath?: string;
}

/**
 * Parse target into server and tool name
 */
function parseTarget(target: string): { server: string; tool: string } {
  const slashIndex = target.indexOf('/');
  if (slashIndex === -1) {
    throw new Error(formatCliError(invalidTargetError(target)));
  }
  return {
    server: target.substring(0, slashIndex),
    tool: target.substring(slashIndex + 1),
  };
}

/**
 * Parse JSON arguments from string or stdin
 */
async function parseArgs(
  argsString?: string,
): Promise<Record<string, unknown>> {
  let jsonString: string;

  if (argsString) {
    jsonString = argsString;
  } else if (!process.stdin.isTTY) {
    // Read from stdin with timeout - use timer cleanup to prevent memory leak
    const timeoutMs = getTimeoutMs();
    const chunks: Buffer[] = [];
    let timeoutId: ReturnType<typeof setTimeout> | undefined;

    const readPromise = (async () => {
      for await (const chunk of process.stdin) {
        chunks.push(chunk);
      }
      return Buffer.concat(chunks).toString('utf-8').trim();
    })();

    const timeoutPromise = new Promise<string>((_, reject) => {
      timeoutId = setTimeout(
        () => reject(new Error(`stdin read timed out after ${timeoutMs}ms`)),
        timeoutMs,
      );
    });

    try {
      jsonString = await Promise.race([readPromise, timeoutPromise]);
    } finally {
      if (timeoutId) clearTimeout(timeoutId);
    }
  } else {
    // No arguments provided
    return {};
  }

  if (!jsonString) {
    return {};
  }

  try {
    return JSON.parse(jsonString);
  } catch (e) {
    throw new Error(
      formatCliError(invalidJsonArgsError(jsonString, (e as Error).message)),
    );
  }
}

/**
 * Execute the call command
 */
export async function callCommand(options: CallOptions): Promise<void> {
  let config: McpServersConfig;

  try {
    config = await loadConfig(options.configPath);
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  let serverName: string;
  let toolName: string;

  try {
    const parsed = parseTarget(options.target);
    serverName = parsed.server;
    toolName = parsed.tool;
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  let serverConfig: ServerConfig;
  try {
    serverConfig = getServerConfig(config, serverName);
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  let args: Record<string, unknown>;
  try {
    args = await parseArgs(options.args);
  } catch (error) {
    console.error((error as Error).message);
    process.exit(ErrorCode.CLIENT_ERROR);
  }

  let connection: McpConnection;

  try {
    connection = await getConnection(serverName, serverConfig);
  } catch (error) {
    console.error(
      formatCliError(
        serverConnectionError(serverName, (error as Error).message),
      ),
    );
    process.exit(ErrorCode.NETWORK_ERROR);
  }

  try {
    const result = await connection.callTool(toolName, args);

    // Extract text content from MCP response for CLI-friendly output
    // Uses formatToolResult which extracts text from MCP content array
    console.log(formatToolResult(result));
  } catch (error) {
    // Try to get available tools for better error message
    let availableTools: string[] | undefined;
    try {
      const tools = await connection.listTools();
      availableTools = tools.map((t) => t.name);
    } catch {
      // Ignore - we'll show error without tool list
    }

    const errMsg = (error as Error).message;
    // Check if it's a "tool not found" type error
    if (errMsg.includes('not found') || errMsg.includes('unknown tool')) {
      console.error(
        formatCliError(toolNotFoundError(toolName, serverName, availableTools)),
      );
    } else {
      console.error(
        formatCliError(toolExecutionError(toolName, serverName, errMsg)),
      );
    }
    process.exit(ErrorCode.SERVER_ERROR);
  } finally {
    await safeClose(connection.close);
  }
}

File: src/commands/call.rs
use crate::errors::CliError;
use std::io::IsTerminal;
use std::time::Duration;
use tokio::io::AsyncReadExt;

pub struct CallOptions {
    pub target: String,       // "server/tool"
    pub args: Option<String>, // JSON arguments
    pub config_path: Option<String>,
}

fn parse_target(target: &str) -> Result<(String, String), CliError> {
    let slash_index = target.find('/');
    if let Some(idx) = slash_index {
        Ok((target[..idx].to_string(), target[idx + 1..].to_string()))
    } else {
        Err(crate::errors::invalid_target_error(target))
    }
}

async fn parse_args(args_string: Option<&str>) -> Result<serde_json::Value, CliError> {
    let json_string = if let Some(args) = args_string {
        args.to_string()
    } else if !std::io::stdin().is_terminal() {
        let mut timeout_ms = crate::config::get_timeout_ms();
        if std::env::var("ANTIGRAVITY_AGENT").is_ok() {
            timeout_ms = 50;
        }
        let mut buffer = Vec::new();
        let mut stdin = tokio::io::stdin();

        let read_fut = stdin.read_to_end(&mut buffer);
        match tokio::time::timeout(Duration::from_millis(timeout_ms), read_fut).await {
            Ok(Ok(_)) => String::from_utf8_lossy(&buffer).trim().to_string(),
            Ok(Err(e)) => {
                return Err(crate::errors::CliError {
                    code: crate::errors::ErrorCode::ClientError,
                    error_type: "STDIN_READ_FAILED".to_string(),
                    message: format!("Failed to read from stdin: {}", e),
                    details: None,
                    suggestion: None,
                });
            }
            Err(_) => {
                return Err(crate::errors::CliError {
                    code: crate::errors::ErrorCode::ClientError,
                    error_type: "STDIN_TIMEOUT".to_string(),
                    message: format!("stdin read timed out after {}ms", timeout_ms),
                    details: None,
                    suggestion: None,
                });
            }
        }
    } else {
        return Ok(serde_json::json!({}));
    };

    if json_string.is_empty() {
        return Ok(serde_json::json!({}));
    }

    serde_json::from_str(&json_string)
        .map_err(|e| crate::errors::invalid_json_args_error(&json_string, Some(&e.to_string())))
}

pub async fn call_command(options: CallOptions) -> Result<(), CliError> {
    let config = crate::config::load_config(options.config_path.as_deref())?;

    let (server_name, tool_name) = parse_target(&options.target)?;

    let server_config = crate::config::get_server_config(&config, &server_name)?;

    let args = parse_args(options.args.as_deref()).await?;

    let connection = crate::client::get_connection(&server_name, &server_config)
        .await
        .map_err(|e| crate::errors::server_connection_error(&server_name, &e.message))?;

    match connection.call_tool(&tool_name, args).await {
        Ok(result) => {
            println!("{}", crate::output::format_tool_result(&result));
            let _ = connection.close().await;
            Ok(())
        }
        Err(error) => {
            let mut available_tools = None;
            if let Ok(tools) = connection.list_tools().await {
                available_tools = Some(tools.into_iter().map(|t| t.name).collect::<Vec<String>>());
            }

            let _ = connection.close().await;

            let err_msg = error.message;
            let err = if err_msg.contains("not found") || err_msg.contains("unknown tool") {
                crate::errors::tool_not_found_error(
                    &tool_name,
                    &server_name,
                    available_tools.as_deref(),
                )
            } else {
                crate::errors::tool_execution_error(&tool_name, &server_name, &err_msg)
            };

            Err(err)
        }
    }
}
